### PR TITLE
Update to JupyterLab 4.0 alpha

### LINKED ISSUE #299: Exiting terminal doesn't close window tab
### Problem

When exiting a terminal by entering `exit`, it would be nice for the window tab to close as well.
Currently, the user is left with an empty page:

image

### Proposed Solution

Maybe use:
```javascript
window.close();
```

### Additional context

That would match JupyterLab's behavior (for which the tab is closed).

### LINKED ISSUE #303: Update to JupyterLab 4.0
### Problem

As proposed in https://github.com/jupyter/enhancement-proposals/pull/79, RetroLab is meant to become Notebook v7.

The JEP proposes to ship have Notebook v7 and JupyterLab 4 work well together, so users can seamlessly switch between the two interfaces.

So we should soon start looking into updating to the latest `4.0` pre-release of JupyterLab.

### Proposed Solution

We update the `@jupyterlab` packages to `4.0.0-alpha.x` (or newer), and the Python package to `4.0.0ax` (or newer).

### PR DESCRIPTION
**Pull request by jtpio**
Fixes #303 
Fixes #299

To start checking for breaking changes, and the new packages / plugins to add to the build.

- [x] Update the `@jupyterlab` packages to `next`
- [x] Update the `@lumino` packages to `latest`
- [x] Fix plugins failing to activate
- [x] Fix notebook toolbar items not displayed

Keeping as a draft for now.

### CODE REVIEW

**Comment by jtpio** [comment]
cc @davidbrochart the build seems to be passing on CI

**Comment by davidbrochart** [comment]
I tried `pip install -e .` with this branch and got:
```
    lerna ERR! yarn run build exited 1 in '@retrolab/lab-extension'
    lerna ERR! yarn run build stdout:
    $ jlpm run build:lib && jlpm run build:labextension:dev
    $ tsc
    src/index.ts(67,5): error TS2322: Type 'Token<IRetroShell>' is not assignable to type 'Token<any>'.
      Types have separate declarations of a private property '_tokenStructuralPropertyT'.
```

**Comment by jtpio** [comment]
Strange. This seems to be building fine locally and also on Binder, which installs from source as follows:

https://github.com/jupyterlab/retrolab/blob/da0975036c27a4a3caff3ce1088a0fefc01574b2/binder/postBuild#L1-L8

image

**Comment by jtpio** [comment]
There is now a `0.3.x` branch created from the `0.3.16` release commit: https://github.com/jupyterlab/retrolab/tree/0.3.x

So we can move forward with this PR and start targeting JupyterLab 4 on `main`.

**Comment by jtpio** [comment]
Still left to do: show the notebook toolbar items, which might be related to this upstream PR: https://github.com/jupyterlab/jupyterlab/pull/10469

image

**Comment by davidbrochart** [comment]
I was trying to set the option to [not close terminals on exit](https://github.com/jupyterlab/jupyterlab/pull/11637), but adding `terminal.content.setOption("closeOnExit", false);` somewhere [here](https://github.com/jupyterlab/retrolab/blob/34702873898b0d2f08c717d7167aec92fc439265/packages/terminal-extension/src/index.ts#L63) doesn't seem to work. Any idea?

**Comment by jtpio** [comment]
@davidbrochart I wonder if this should be set here instead:

https://github.com/jupyterlab/retrolab/blob/34702873898b0d2f08c717d7167aec92fc439265/packages/terminal-extension/src/index.ts#L40

iirc the other plugin handles opening in a new browser tab.

**Comment by jtpio** [comment]
> Still left to do: show the notebook toolbar items, which might be related to this upstream PR: [jupyterlab/jupyterlab#10469](https://github.com/jupyterlab/jupyterlab/pull/10469)

This one looks like a race condition. Adding an artifical delay such as following seems lets the toolbar be correctly populated:

```diff
diff --git a/packages/application-extension/src/index.ts b/packages/application-extension/src/index.ts
index 109fd19..866570b 100644
--- a/packages/application-extension/src/index.ts
+++ b/packages/application-extension/src/index.ts
@@ -181,9 +181,11 @@ const opener: JupyterFrontEndPlugin<void> = {
         app.restored.then(() => {
           // TODO: get factory from file type instead?
           if (ext === '.ipynb') {
-            docManager.open(file, NOTEBOOK_FACTORY, undefined, {
-              ref: '_noref'
-            });
+            setTimeout(() => {
+              docManager.open(file, NOTEBOOK_FACTORY, undefined, {
+                ref: '_noref'
+              });
+            }, 1000);
           } else {
             docManager.open(file, EDITOR_FACTORY, undefined, {
               ref: '_noref'
```

https://user-images.githubusercontent.com/591645/150326260-c99bcc93-49fc-4cfa-ba5d-9c1fc4327c3a.mp4

**Comment by davidbrochart** [comment]
Thanks, that was it!
I opened https://github.com/jtpio/retrolab/pull/4.

**Comment by jtpio** [comment]
Nice thanks @davidbrochart!

**Comment by jtpio** [comment]
> Adding an artificial delay such as following seems lets the toolbar be correctly populated

cc @fcollonval who might now if this is related to the logic in https://github.com/jupyterlab/jupyterlab/pull/10469

**Comment by jtpio** [comment]
Using the following trick like in the `toolbar.spec.ts` upstream tests seems to do it:

https://github.com/jupyterlab/jupyterlab/blob/50fa0047d63287005a2c3b1d8f5bfedfe56cde7e/packages/apputils/test/toolbar.spec.ts#L304-L308

```ts
await settingRegistry?.load('@jupyterlab/notebook-extension:panel');
await Promise.resolve();
```

**Comment by jtpio** [comment]
This is what the user menu dropdown looks like in RetroLab when RTC is enabled:

image

File: packages/application-extension/src/index.ts [hunk 1]
```diff
@@ -178,9 +180,12 @@ const opener: JupyterFrontEndPlugin<void> = {
 
         const file = decodeURIComponent(path);
         const ext = PathExt.extname(file);
-        app.restored.then(() => {
+        app.restored.then(async () => {
           // TODO: get factory from file type instead?
           if (ext === '.ipynb') {
+            // TODO: fix upstream?
+            await settingRegistry?.load('@jupyterlab/notebook-extension:panel');
+            await Promise.resolve();
```
Comment (jtpio) [inline packages/application-extension/src/index.ts:188]: We should investigate if it is possible to fix this upstream as this is very brittle.

**Comment by jtpio** [comment]
Merging that one since CI is now green.

Will make a `0.4.0a0` prerelease with this right after.

**Comment by jtpio** [comment]
`0.4.0a0` is out: https://github.com/jupyterlab/retrolab/releases/tag/v0.4.0a0

File: packages/application-extension/src/index.ts (hunk 1 above)
Comment (fcollonval) [reply packages/application-extension/src/index.ts:188]: A toolbar definition may indeed not be fully known by the time the open is triggered as some plugins may not yet be loaded or the treatment of the settings may not be complete yet. Hence adding some sort of delay does the trick.

**Pull request merged**